File: crypto/envelope.py
import json
import base64
import uuid
import time

from crypto.pqc import (
    kem_encapsulate, kem_decapsulate,
    sign_message, verify_signature,
    hybrid_kem_encapsulate, hybrid_kem_decapsulate,
)
from crypto.symmetric import encrypt, decrypt


ENVELOPE_VERSION = 2


def _b64(data: bytes) -> str:
    return base64.b64encode(data).decode("ascii")


def _unb64(data: str) -> bytes:
    return base64.b64decode(data)


def _lp(data: bytes) -> bytes:
    return len(data).to_bytes(4, "big") + data


def _lp_str(s: str) -> bytes:
    return _lp(s.encode("utf-8"))


def _build_context_aad(sender_id: str, recipient_id: str, subject: str,
                       message_id: str, timestamp: int) -> bytes:
    return (
        _lp_str(sender_id)
        + _lp_str(recipient_id)
        + _lp_str(subject)
        + _lp_str(message_id)
        + _lp(timestamp.to_bytes(8, "big"))
    )


def _build_signed_payload_v2(
    sender_id: str, recipient_id: str, subject: str,
    message_id: str, timestamp: int,
    x25519_ephem_pk: bytes, kem_ct: bytes,
    nonce: bytes, ciphertext: bytes, tag: bytes,
) -> bytes:
    return (
        _lp_str(sender_id)
        + _lp_str(recipient_id)
        + _lp_str(subject)
        + _lp_str(message_id)
        + _lp(timestamp.to_bytes(8, "big"))
        + _lp(x25519_ephem_pk)
        + _lp(kem_ct)
        + _lp(nonce)
        + _lp(ciphertext)
        + _lp(tag)
    )


def _build_signed_payload_v1(kem_ciphertext: bytes, nonce: bytes,
                             ciphertext: bytes, tag: bytes) -> bytes:
    return kem_ciphertext + nonce + ciphertext + tag


def seal_envelope(
    plaintext: bytes,
    recipient_kem_pk: bytes,
    sender_sign_sk: bytes,
    sender_id: str,
    recipient_id: str,
    *,
    recipient_x25519_pk: bytes,
    subject: str = "",
    message_id: str | None = None,
) -> str:
    if message_id is None:
        message_id = str(uuid.uuid4())
    timestamp = int(time.time())

    kem_ct, x25519_ephem_pk, session_key = hybrid_kem_encapsulate(
        recipient_kem_pk, recipient_x25519_pk,
    )

    aad = _build_context_aad(sender_id, recipient_id, subject, message_id, timestamp)
    nonce, ciphertext, tag = encrypt(session_key, plaintext, aad=aad)

    signed_payload = _build_signed_payload_v2(
        sender_id, recipient_id, subject, message_id, timestamp,
        x25519_ephem_pk, kem_ct, nonce, ciphertext, tag,
    )
    signature = sign_message(sender_sign_sk, signed_payload)

    envelope = {
        "version": ENVELOPE_VERSION,
        "kem": "X25519+ML-KEM-768",
        "sig": "ML-DSA-65",
        "sym": "AES-256-GCM",
        "kdf": "HKDF-SHA256",
        "sender_id": sender_id,
        "recipient_id": recipient_id,
        "message_id": message_id,
        "timestamp": timestamp,
        "subject": subject,
        "x25519_ephemeral_pk": _b64(x25519_ephem_pk),
        "kem_ciphertext": _b64(kem_ct),
        "nonce": _b64(nonce),
        "ciphertext": _b64(ciphertext),
        "tag": _b64(tag),
        "signature": _b64(signature),
    }

    return json.dumps(envelope)


def open_envelope(
    envelope_json: str,
    recipient_kem_sk: bytes,
    recipient_x25519_sk: bytes | None = None,
    sender_verify_key: bytes | None = None,
) -> bytes:
    envelope = json.loads(envelope_json)
    version = envelope.get("version", 1)

    if version == 1:
        return _open_v1(envelope, recipient_kem_sk)
    if version == 2:
        if recipient_x25519_sk is None or sender_verify_key is None:
            raise ValueError("V2 envelopes require recipient_x25519_sk and sender_verify_key")
        return _open_v2(envelope, recipient_kem_sk, recipient_x25519_sk, sender_verify_key)

    raise ValueError(f"Unsupported envelope version: {version}")


def _open_v2(envelope: dict, recipient_kem_sk: bytes,
             recipient_x25519_sk: bytes, sender_verify_key: bytes) -> bytes:
    kem_ct = _unb64(envelope["kem_ciphertext"])
    x25519_ephem_pk = _unb64(envelope["x25519_ephemeral_pk"])
    nonce = _unb64(envelope["nonce"])
    ciphertext = _unb64(envelope["ciphertext"])
    tag = _unb64(envelope["tag"])
    signature = _unb64(envelope["signature"])

    sender_id = envelope["sender_id"]
    recipient_id = envelope["recipient_id"]
    subject = envelope.get("subject", "")
    message_id = envelope["message_id"]
    timestamp = envelope["timestamp"]

    signed_payload = _build_signed_payload_v2(
        sender_id, recipient_id, subject, message_id, timestamp,
        x25519_ephem_pk, kem_ct, nonce, ciphertext, tag,
    )
    if not verify_signature(sender_verify_key, signed_payload, signature):
        raise ValueError("Signature verification failed")

    session_key = hybrid_kem_decapsulate(
        recipient_kem_sk, recipient_x25519_sk, kem_ct, x25519_ephem_pk,
    )

    aad = _build_context_aad(sender_id, recipient_id, subject, message_id, timestamp)
    return decrypt(session_key, nonce, ciphertext, tag, aad=aad)


def _open_v1(envelope: dict, recipient_kem_sk: bytes) -> bytes:
    kem_ciphertext = _unb64(envelope["kem_ciphertext"])
    nonce = _unb64(envelope["nonce"])
    ciphertext = _unb64(envelope["ciphertext"])
    tag = _unb64(envelope["tag"])
    signature = _unb64(envelope["signature"])
    sender_verify_key = _unb64(envelope["sender_verify_key"])

    signed_payload = _build_signed_payload_v1(kem_ciphertext, nonce, ciphertext, tag)
    if not verify_signature(sender_verify_key, signed_payload, signature):
        raise ValueError("Signature verification failed — message may be tampered")

    shared_secret = kem_decapsulate(recipient_kem_sk, kem_ciphertext)
    return decrypt(shared_secret, nonce, ciphertext, tag)

File: frontend/src/crypto/envelope.ts
import { sign, verify } from './pqc';
import { encrypt, decrypt } from './symmetric';
import { hybridKemEncapsulate, hybridKemDecapsulate } from './hybrid';
import type { EnvelopeV2 } from './types';

const ENVELOPE_VERSION = 2;
const encoder = new TextEncoder();
const decoder = new TextDecoder();

function b64Encode(data: Uint8Array): string {
  let binary = '';
  for (let i = 0; i < data.length; i++) binary += String.fromCharCode(data[i]);
  return btoa(binary);
}

function b64Decode(s: string): Uint8Array {
  const binary = atob(s);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

function lp(data: Uint8Array): Uint8Array {
  const len = new DataView(new ArrayBuffer(4));
  len.setUint32(0, data.length);
  const out = new Uint8Array(4 + data.length);
  out.set(new Uint8Array(len.buffer));
  out.set(data, 4);
  return out;
}

function lpStr(s: string): Uint8Array {
  return lp(encoder.encode(s));
}

function concatAll(...arrays: Uint8Array[]): Uint8Array {
  let total = 0;
  for (const a of arrays) total += a.length;
  const out = new Uint8Array(total);
  let offset = 0;
  for (const a of arrays) {
    out.set(a, offset);
    offset += a.length;
  }
  return out;
}

function timestampBytes(ts: number): Uint8Array {
  const buf = new DataView(new ArrayBuffer(8));
  buf.setBigUint64(0, BigInt(ts));
  return new Uint8Array(buf.buffer);
}

function buildContextAad(
  senderId: string,
  recipientId: string,
  subject: string,
  messageId: string,
  timestamp: number,
): Uint8Array {
  return concatAll(
    lpStr(senderId),
    lpStr(recipientId),
    lpStr(subject),
    lpStr(messageId),
    lp(timestampBytes(timestamp)),
  );
}

function buildSignedPayloadV2(
  senderId: string,
  recipientId: string,
  subject: string,
  messageId: string,
  timestamp: number,
  x25519EphemPk: Uint8Array,
  kemCt: Uint8Array,
  nonce: Uint8Array,
  ciphertext: Uint8Array,
  tag: Uint8Array,
): Uint8Array {
  return concatAll(
    lpStr(senderId),
    lpStr(recipientId),
    lpStr(subject),
    lpStr(messageId),
    lp(timestampBytes(timestamp)),
    lp(x25519EphemPk),
    lp(kemCt),
    lp(nonce),
    lp(ciphertext),
    lp(tag),
  );
}

function generateUUID(): string {
  return crypto.randomUUID();
}

export async function sealEnvelope(
  plaintext: Uint8Array,
  recipientKemPk: Uint8Array,
  senderSignSk: Uint8Array,
  senderId: string,
  recipientId: string,
  recipientX25519Pk: Uint8Array,
  subject = '',
  messageId?: string,
): Promise<string> {
  if (!messageId) messageId = generateUUID();
  const timestamp = Math.floor(Date.now() / 1000);

  const { kemCt, ephemeralPk, sessionKey } = await hybridKemEncapsulate(
    recipientKemPk,
    recipientX25519Pk,
  );

  const aad = buildContextAad(senderId, recipientId, subject, messageId, timestamp);
  const { nonce, ciphertext, tag } = await encrypt(sessionKey, plaintext, aad);

  const signedPayload = buildSignedPayloadV2(
    senderId, recipientId, subject, messageId, timestamp,
    ephemeralPk, kemCt, nonce, ciphertext, tag,
  );
  const signature = sign(senderSignSk, signedPayload);

  const envelope: EnvelopeV2 = {
    version: ENVELOPE_VERSION,
    kem: 'X25519+ML-KEM-768',
    sig: 'ML-DSA-65',
    sym: 'AES-256-GCM',
    kdf: 'HKDF-SHA256',
    sender_id: senderId,
    recipient_id: recipientId,
    message_id: messageId,
    timestamp,
    subject,
    x25519_ephemeral_pk: b64Encode(ephemeralPk),
    kem_ciphertext: b64Encode(kemCt),
    nonce: b64Encode(nonce),
    ciphertext: b64Encode(ciphertext),
    tag: b64Encode(tag),
    signature: b64Encode(signature),
  };

  return JSON.stringify(envelope);
}

export async function openEnvelope(
  envelopeJson: string,
  recipientKemSk: Uint8Array,
  recipientX25519Sk: Uint8Array,
  senderVerifyKey: Uint8Array,
): Promise<Uint8Array> {
  const envelope = JSON.parse(envelopeJson) as EnvelopeV2;

  if (envelope.version !== 2) {
    throw new Error(`Unsupported envelope version: ${envelope.version}`);
  }

  const kemCt = b64Decode(envelope.kem_ciphertext);
  const x25519EphemPk = b64Decode(envelope.x25519_ephemeral_pk);
  const nonce = b64Decode(envelope.nonce);
  const ciphertext = b64Decode(envelope.ciphertext);
  const tag = b64Decode(envelope.tag);
  const signature = b64Decode(envelope.signature);

  const { sender_id, recipient_id, message_id, timestamp } = envelope;
  const subject = envelope.subject ?? '';

  const signedPayload = buildSignedPayloadV2(
    sender_id, recipient_id, subject, message_id, timestamp,
    x25519EphemPk, kemCt, nonce, ciphertext, tag,
  );

  if (!verify(senderVerifyKey, signedPayload, signature)) {
    throw new Error('Signature verification failed');
  }

  const sessionKey = await hybridKemDecapsulate(
    recipientKemSk, recipientX25519Sk, kemCt, x25519EphemPk,
  );

  const aad = buildContextAad(sender_id, recipient_id, subject, message_id, timestamp);
  return decrypt(sessionKey, nonce, ciphertext, tag, aad);
}

export function decodeEnvelopeText(plaintext: Uint8Array): string {
  return decoder.decode(plaintext);
}
